Remove the global "selected room" from online screens

smoogipoo opened the pull request:
The global `Bindable<Room>` has caused a lot of headaches, such as requiring a lease to be taken out in `LoungeSubScreen` (which I've left, but is probably not needed anymore) and that the caching of room components was hidden.

In this PR;
- Selected room moved from `OnlinePlayScreen` to `LoungeSubScreen`.
- `DrawableRoom` and `RoomSubScreen` cache the `Room` via `CachedModelDependencyContainer` for use by sub-components.
  - Removing this is a much larger change, but can be done in a future PR. I'm not 100% on this, but we can discuss afterwards.
- Components like `RoomSettingsOverlay` that need the specific `Room` now receive the `Room` directly.
- Backgrounds have been reworked, and now use the same `BackgroundScreenStack` that's used in top-level `OsuGame` screens.
  - This is the bulk of changes in this PR. Unfortunately it's quite a big one because backgrounds were a global concept before.

One thing to note is that I've removed the `currentRoom == null` check in `MultiplayerMatchSubScreen.LoadComplete()`. This is a bit hard to replace, but it should be added back following the splitting out of `RoomSettingsOverlay` into a separate screen before `RoomSubScreen` is entered.  
I don't believe this is a crucial component of this screen...

The transition to song select is slightly jarring because I haven't changed the background it uses, so it still does the x-offset. I think that can be adjusted in a separate PR.

Preview: https://drive.google.com/file/d/1pkb5L7koPx-J5dia8DgS8kMMaogMSzAl/view

smoogipoo committed:
Pass room into RoomSettingsOverlay

smoogipoo committed:
Pass room into SelectionPollingComponent

smoogipoo committed:
Make RoomsContainer/DrawableRoom not resolve via DI

smoogipoo committed:
Remove unnecessary resolve into ListingPollingComponent

smoogipoo committed:
Add dependency container to RoomSubScreen

smoogipoo committed:
Disable condition for the time being

smoogipoo committed:
Use BackgroundScreen for multiplayer backgrounds

smoogipoo committed:
Initial rework of backgrounds

smoogipoo committed:
Merge branch 'master' into remove-current-room

smoogipoo committed:
Move screenstack up a level

smoogipoo committed:
Set a default background

smoogipoo committed:
Fix background overflows

smoogipoo committed:
Initial room background implementation

smoogipoo committed:
Block lounge background screen from exiting

smoogipoo committed:
Make DrawableMatchRoom background load instantly

smoogipoo committed:
Merge branch 'master' into remove-current-room

smoogipoo committed:
Remove selectedRoom from OnlinePlayScreen

smoogipoo committed:
Simplify lounge implementation

smoogipoo committed:
Fix initial display in room background

smoogipoo committed:
Cleanups

peppy commented:
Actually with the background handling moved as it is, the horizontal movement can probably be added to the other multiplayer screens to match. I'll take a look at this during review.

Also, test failures look relevant.

peppy commented:
Before going through this in detail, I do find it a bit unfortunate that the default background shows up after entering the room briefly. On `master` it was empty instead. Probably something we can fix down the line?

https://user-images.githubusercontent.com/191335/130390026-a5708e1c-f4db-48e6-ace2-22240790b404.mp4

peppy commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -0,0 +1,101 @@
+        {
+            base.Update();
+
+            // This is a static screen, so override the scale set in base.Update(), but also the scale set by the screen stack.
+            Scale = new Vector2(1f / BackgroundScreenStack.BACKGROUND_SCALE);
```
Where can I see this taking effect? Even when reading the inline comment and commenting this out I can't seem to notice the issue (or figure what this is trying to achieve).

smoogipoo replied:
Comment this out and open `TestSceneMultiplayer` at 10% speed, you'll see the background poking out from the top into the waves.

peppy replied:
Do you mean this?

![20210823 134743 (dotnet)](https://user-images.githubusercontent.com/191335/130391352-72b82929-6885-473b-9da6-03860fa1bb45.png)

It seems to happen even with the scale applied, just a bit less.

peppy replied:
This is how it looks on your branch currently in game testing:

https://user-images.githubusercontent.com/191335/130391484-623bb803-f177-4fd5-bbda-de55226912e1.mp4

It gets less noticeable with taller windows, it would seem.

smoogipoo replied:
That looks like you've just set the scale to `Vector2.One`. There are two places where scale is applied - one in `BackgroundScreenStack` with value `1.06f` which is what I assume you're seeing there, and a second time in this class' `base.Update()` which is much larger than that.

I don't know where you're getting your results from, this is how it looks for me:
![image](https://user-images.githubusercontent.com/1329837/130391760-cd19c44a-6dfe-4889-80d5-bed01694b956.png)
![image](https://user-images.githubusercontent.com/1329837/130391785-a3b102fa-eeae-4217-a9e7-378f7d005701.png)

peppy replied:
I'm just testing on your branch with no changes, running the game using the `osu.Desktop` project.

Here's two window sizes:

## Tall

![20210823 135618 (dotnet)](https://user-images.githubusercontent.com/191335/130392456-73f82d97-1fd0-4f89-a27b-7f8b7ce931f5.png)

## Wide (as per video)

![20210823 135651 (dotnet)](https://user-images.githubusercontent.com/191335/130392502-3ccaf308-d230-4e3e-baaa-4c86e02948fb.png)

smoogipoo replied:
I can't repro this. I've tested both visual tests and in-game at both resolutions. It's either something on your end, or something you need to investigate.

peppy replied:
As discussed, this is because you didn't have parallax enabled. I've moved the gradient to the background itself to fix this issue in a more global way. Note that with parallax enabled, this results in the background screen extending further above the actual lounge screens than expected. In an optimal world we would be able to turn on masking to avoid this, but due to the aforementioned bug this is not done yet.

Probably something to reconsider if it ever becomes an issue in the future. Right now it's almost indiscernible (if you look closely at the wave container effect you'll notice that it's not as visible as usual, but it all happens very fast).

smoogipoo commented:
> I do find it a bit unfortunate that the default background shows up after entering the room briefly

So a few things:
1. I'm fairly certain the current background is incorrect and doesn't follow the current beatmap.
2. This is caused by the removal of the beatmap, which is the same reason that the beatmap panel also disappears briefly.
3. This should only happen when creating a room, not joining.

And yes, these things should be fixed in post, following other refactorings like splitting out room creation into a screen.

peppy commented on osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs:
```diff
@@ -34,11 +34,10 @@ public class DrawableLoungeRoom : DrawableRoom, IFilterable, IHasContextMenu, IH
 
-        [Resolved(canBeNull: true)]
-        private LoungeSubScreen lounge { get; set; }
+        public readonly Bindable<Room> SelectedRoom = new Bindable<Room>();
```
Feels a bit weird to still have this here. Might work better to move the "is selected" logic up one level (to `RoomsContainer`) and change this to 

```suggestion
        public readonly BindableBool IsSelected = new BindableBool(true);
```

Right now, the component can't be interacted with without `SelectedRoom` both being bound and being set to `true`, which is a bit of a hidden dependency.

smoogipoo replied:
I didn't want to do this in order to to keep the changes small.

peppy replied:
I guess this line saves the flow from being completely broken:

https://github.com/ppy/osu/blob/d304e283e4a24506923f48097b84494086ef38d6/osu.Game/Screens/OnlinePlay/Lounge/DrawableLoungeRoom.cs#L162

Let's leave it for a follow-up change.

smoogipoo committed:
Fix test failure

peppy committed:
Remove background scale altogether

I'm not sure why this is required. Seems like something which was meant
to exist to handle parallax, but that is already handled elsewhere now.

peppy committed:
Move gradient to be part of the background for now
